docs/api/v1.md

```diff
@@ -0,0 +1,117 @@
+# Spice API
+## Version `v1`
+
+All routes are prefixed with `/api/v1/`
+
+Example: `GET localhost:1234/api/v1/processes`
+
+## Host System Info
+
+### `GET /filesystem/:path*`
+Gets the file(s) within the given path
+
+Path parameters:
+ - `path` {string}: URL encoded path to a file or directory.  If empty, refers to the root directory.
+
+Responses:
+ - `200`:
+    - Returns {[File](#File)}
+ - `400`:
+    - Returns {[Error](#Error)}: Invalid path.
+ - `403`: 
+    - Returns {[Error](#Error)}: Insufficient privileges to open file.
+ - `404`: 
+    - Returns {[Error](#Error)}: File doesn't exit.
+
+### `GET /processes`
+Gets the list of processes running on the host machine.
+
+Responses:
+ - `200`:
+    - Returns {[Process](#Process)[]}
+ - `403`:
+    - Returns {[Error](#Error)}: Insufficient privileges to complete operation.
+
+## Debug
+All `/debug` routes are stateful--their responses are based on the currently attached process/executable.
+If the debugger is not attached, they respond with `40somethin`.
+
+### `POST /debug/attach/pid/:pid`
+Returns status of attached pid, execution id of current function.
+
+### `POST /debug/attach/bin/:path*`
+
+### `GET /debug`
+Returns information about the attached process or executable.
+
+### `GET /debug/functions`
+Returns a list of functions in the attached process.
+
+### `GET /debug/functions/:function`
+Returns source code and input fields for this function.
+
+### `PUT /debug/functions/:function/breakpoints`
+Sets a breakpoint on this function.
+
+### `DELETE /debug/functions/:function/breakpoints`
+Removes breakpoint on this function.
+
+### `GET /debug/breakpoints`
+Lists breakpoints.
+
+### `POST /debug/functions/:function/execute`
+Executes the function with parameters provided in POST body.
+Returns an [execution id](#execution).
+
+### `GET/debug/executions`
+Returns list of all executions.
+
+### `GET /debug/executions/:executionId`
+Returns execution status (function, processing/done, execution time).
+
+### `GET /debug/executions/:executionId/trace`
```

Comment:
There were a few minor reasons for this design, but it's very much up in the air. Most of the reasons are moot in the local environment, but we might as well plan for remote and multi-user debugging.

1. Minimize call state
I can imagine debugging a running program would involve some kind of `/debug/continue` call, which executes until the end of the next breakpoint function. If we stream back the trace in the response of this call, we can end up in some nasty situations where the function executed on the remote, but the client never gets the response, or multiple users both call "continue" at the same time. We can reasonably handle these situations in other ways, but in my option making more "idempotent" endpoints is the simplest solution. If these issues do come up, we have a short history of traces on the server to fall back on.

2. Separate streaming endpoints
Whenever this "continue" endpoint is called, we need to send the client information about the function/file so it can display the source code correctly; then in the same response we stream the trace. I'd prefer to have streaming data in separate API calls to simplify implementation (because of the way libraries are typically structured, it's a lot easier to get some function metadata as a JSON blob, then in a separate request handle streaming multiple lines of JSON objects, rather than having to parse some JSON, then handle extra streaming data in the same request).

3. Separation of concerns
We already need to provide "filesystem" endpoints to retrieve source code, so it's less complex and less work to allow the client to reuse those endpoints, rather than streaming back the source code in the `/debug/continue` call.

4. Pagination
We can provide query parameters to the `GET trace` endpoint to allow filtering of results, potentially using queries that filter by level of recursion/loop, etc. We might want to just do this on the client instead, as you said.

These are pretty small concerns, but I think this API has some tangible benefits. On low-memory machines we can limit the "history" to 1 or 2 traces.